creator/ingest_runs/tasks.py

```diff
@@ -1,13 +1,109 @@
+    """
+    Run ingest process for a given set of files.
+    """
+    ingest_run = get_ingest_run(ingest_run_id)
+    logging.info(f"Preparing ingest run {ingest_run.pk} for processing.")
+
+    # Check if any IngestRuns with the same input hash exist. If so, cancel
+    # them and delete the newly created IngestRun, which never gets started.
+    # If not, start a new IngestRun.
+    if cancel_duplicate_ingest_runs(ingest_run):
+        ingest_run.delete()
+    else:
+        # Update run state so we know it's running
+        logging.info(f"Start ingest {ingest_run.id}")
+        ingest_run.start()
+        ingest_run.save()
+
+        # Check that the files are genomic workflow manifests
+        versions = ingest_run.versions.all()
+        are_gwo = (check_gwo(version) for version in versions)
+        if all(are_gwo):
+            logging.info(
+                "All files in the ingest run are genomic workflow manifests."
+            )
+            ingest_genomic_workflow_output_manifests(ingest_run)
+        else:
+            raise Exception("Unknown file type detected in the ingest run.")
+        logging.info(f"Ingest run {ingest_run.id} complete!")
+        ingest_run.complete()
+        ingest_run.save()
+
+
+def cancel_duplicate_ingest_runs(ingest_run):
+    """
+    Check for other IngestRuns with the same input hash. If they exist, cancel
+    them. Returns a flag indicating whether duplicated IngestRuns were found.
+    """
+    others = IngestRun.objects.filter(
+        state="started",
+        input_hash=ingest_run.input_hash
+    ).all()
+    others_flag = others.count() > 0
 
+    # Cancel them
+    for running_ingest in others:
+        ingest_id = str(running_ingest.id)
+        logging.info(f"Canceling previous run '{ingest_id}'")
+        django_rq.enqueue(cancel_ingest, ingest_id)
 
-def ingest_genomic_workflow_ouput_manifests(versions):
+    return others_flag
+
+
+def check_gwo(version):
+    return version.root_file.file_type == FileType.GWO.value
+
+
+@task("cancel_ingest")
+def cancel_ingest(ingest_run_id=None):
+    logging.info(f"Canceling ingest run {ingest_run_id}...")
+    ingest_run = get_ingest_run(ingest_run_id)
+    ingest_run.cancel()
+    ingest_run.save()
+
+
+def ingest_genomic_workflow_output_manifests(ingest_run):
     # TODO
-    print(f"Begin ingesting genomic workflow manifests: {versions}")
+    print(
+        f"Begin ingesting genomic workflow manifests: "
+        f"{list(ingest_run.versions.all())}"
+    )
+    import time
+    t_end = time.time() + 10
+    progress = ''
+    while time.time() < t_end:
+        time.sleep(1)
+        progress += '.'
+        print(progress)
+    print(f"Finished ingest run {ingest_run.id}")
+
+
+def get_ingest_run(ingest_run_id):
+    """
+    Obtain the IngestRun with id _ingest_run_id_.
+    """
+    try:
+        ingest_run = IngestRun.objects.get(pk=ingest_run_id)
+    except IngestRun.DoesNotExist as err:
+        logging.info(
```

Comment:
Probably want a `ingest_run.fail()` here too

Reply:
I think you could remove this log statement because when the exception is raised it will get logged by the job logger/task decorator and have pretty much the same info in it.

Reply:
If there's an exception then there's no `ingest_run` to use

Reply:
Oh oops, you're right